packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -68,14 +168,54 @@ class LinuxUserData extends UserData {
     this.lines.push(...commands);
   }
 
+  public addOnExitCommands(...commands: string[]) {
+    this.onExitLines.push(...commands);
+  }
+
   public render(): string {
     const shebang = this.props.shebang !== undefined ? this.props.shebang : '#!/bin/bash';
-    return [shebang, ...this.lines].join('\n');
+    return [shebang, ...(this.renderOnExitLines()), ...this.lines].join('\n');
+  }
+
+  public addS3DownloadCommand(params: S3DownloadOptions): string {
+    const s3Path = `s3://${params.bucket.bucketName}/${params.bucketKey}`;
+    const localPath = ( params.localFile && params.localFile.length !== 0 ) ? params.localFile : `/tmp/${ params.bucketKey }`;
+    this.addCommands(
+      `mkdir -p $(dirname '${localPath}')`,
+      `aws s3 cp '${s3Path}' '${localPath}'`
+    );
+
+    return localPath;
+  }
+
+  public addExecuteFileCommand( params: ExecuteFileOptions): void {
+    this.addCommands(
+      `set -e`,
```

Comment:
The download also needs to execute under `set -e`. An easier way to set this is to change the shebang.

Can you change the default shebang to `#!/bin/bash -e` ?

Reply:
Potentially even `#!/bin/bash -ex`

Reply:
I am not sure if we want to actually to modify the shebang like this since I would consider the following:
* -e being a potentially breaking change for users.  Anyone who was already adding -e manually would be fine but everyone else could run into unepected failures.
* -x being a potential security risk since unless customers explicitly disabled it we could be printing secrets in the userdata.

Reply:
That is a fair concern, but by that logic we can never improve any situation. At least `-e` should be the default for 99% of users; people who want/need something else because they do custom error handling should be able to deviate from the default if they want to.

I will take this change myself since I also predict it will be a lot of integ test work.